opennebula/provider.go

```diff
@@ -72,30 +75,57 @@ func Provider() *schema.Provider {
 }
 
 type Configuration struct {
+	OneVersion string
 	Controller *goca.Controller
 	mutex      MutexKV
 }
 
 func providerConfigure(d *schema.ResourceData) (interface{}, error) {
-	one_client := goca.NewDefaultClient(goca.NewConfig(d.Get("username").(string),
-		d.Get("password").(string),
-		d.Get("endpoint").(string)))
 
-	if flow_endpoint, ok := d.GetOk("flow_endpoint"); ok {
-		flow_client := goca.NewDefaultFlowClient(
-			goca.NewFlowConfig(d.Get("username").(string),
-				d.Get("password").(string),
-				flow_endpoint.(string)))
+	username, ok := d.GetOk("username")
+	if !ok {
+		return nil, fmt.Errorf("username should be defined")
+	}
+
+	password, ok := d.GetOk("password")
+	if !ok {
+		return nil, fmt.Errorf("password should be defined")
+	}
+
+	endpoint, ok := d.GetOk("endpoint")
+	if !ok {
+		return nil, fmt.Errorf("endpoint should be defined")
+	}
+
+	oneClient := goca.NewDefaultClient(goca.NewConfig(username.(string),
+		password.(string),
+		endpoint.(string)))
+
+	version, err := goca.NewController(oneClient).SystemVersion()
+	if err != nil {
+		return nil, err
+	}
+
+	log.Printf("[INFO] OpenNebula version: %s", version)
+
+	flowEndpoint, ok := d.GetOk("flow_endpoint")
+	if ok {
```

Comment:
What does happen if not `ok`?

Reply:
The flow configuration is optional in the provider: oneflow is an optional service to configure in OpenNebula.
If we don't provider the flow endpoint then the provider won't be able to work with flow resources (the service resources).

So here, if not ok, the provider will get a controller that isn't able to work with flow resources.